### PR TITLE
Move tensor.index_copy_ 's CPU path to ATen

### LINKED ISSUE #1777: index_copy inconsistent behavior on CPU and GPU
`index_copy` has inconsistent behaviors for the shape of the tensor which is used to copy on CPU and GPU. In following examples, `index_copy` raises an error on CPU due to the invalid shape of `y`, but it works well on GPU for same tensors `x`, `y`, `z`.
```
In [1]: import torch

In [2]: from torch.autograd import Variable

In [3]: x = torch.ones(3,1)

In [4]: y = torch.Tensor([4])

In [5]: z = torch.LongTensor([2])

In [6]: x.index_copy_(0, z, y)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-a1189d076165> in <module>()
----> 1 x.index_copy_(0, z, y)

RuntimeError: cannot select on a vector at /users1/xwgeng/pytorch/torch/lib/TH/generic/THTensor.c:403

In [7]: x = x.cuda()

In [8]: y = y.cuda()

In [9]: z = z.cuda()

In [10]: x.index_copy_(0, z, y)
Out[10]: 

 1
 1
 4
[torch.cuda.FloatTensor of size 3x1 (GPU 0)]
```
The pytorch version is `0.1.12+d7db75c`

### PR DESCRIPTION
**Pull request by zou3519**
Fixes #1777.

Previously, there was a bug where index_copy on CUDA would not error out if the shape of the source slice (in dimension 0) is not equal to the shape of the destination slice (in the specified dimension).
For example,
```
x = torch.zeros(3, 1)
t = torch.arange(3)
index = torch.LongTensor([2, 1, 0])
x.index_copy_(0, index, t) # x has slizes of size (1, 1) in dimension 0 but t has zero-dim slices in dimension 0.
```
would work on CUDA but not on CPU. This PR changes it so that both the CPU and CUDA paths use the same error checking so that the above code will error out on both CPU and CUDA.

To do this, I moved the CPU implementation of index_copy_ to ATen.

### CODE REVIEW

File: aten/src/ATen/native/Indexing.cpp [hunk 1]
```diff
@@ -250,4 +250,50 @@ Tensor & index_put_(Tensor & self, TensorList indices, const Tensor & value) {
...
+  if (self.is_cuda()) {
+    return self._indexCopy_(dim, index, source);
+  }
+
+  auto workingIndices = index.contiguous();
+  int64_t* index_data = reinterpret_cast<int64_t*>(workingIndices.data_ptr());
+  for (int64_t i = 0; i < numIndices; i++) {
+    auto selfSlice = self.select(dim, index_data[i]);
+    auto sourceSlice = source.select(dim, i);
+    selfSlice.copy_(sourceSlice);
```
Comment (fmassa) [inline aten/src/ATen/native/Indexing.cpp]: Quick question: is there a performance regression for 1D arrays? There used to be a dedicated codepath that avoided `select` and `copy`.

Comment (zou3519) [reply aten/src/ATen/native/Indexing.cpp]: Possibly. Let me run some numbers.

Comment (zou3519) [reply aten/src/ATen/native/Indexing.cpp]: Thanks for catching that, @fmassa, the code is 100x slower now :( :( 

In that case, I'll leave index_copy_ in TH.

**Comment by zou3519** [comment]
Closing in favor of a solution that doesn't involve migrating index_copy to aten.

**Pull request closed without merging**